### PR TITLE
Add service for document uploads

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Create document upload job](https://tools.hmcts.net/jira/browse/BPS-717)

### Change description ###

Re-creating #1162 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+            .willReturn(new ZipFileProcessingResult(new byte[]{}, emptyList())); // unit test doesn't care if it's empty
+        willDoNothing().given(documentProcessor).uploadPdfFiles(emptyList(), emptyList());
+
+        // when
+        uploadService.processByContainer(CONTAINER_1, getEnvelopes());
+
+        // then
+        ArgumentCaptor<Event> eventCaptor = ArgumentCaptor.forClass(Event.class);
+        verify(envelopeProcessor, times(1)).handleEvent(any(Envelope.class), eventCaptor.capture());
+        assertThat(eventCaptor.getValue()).isEqualTo(Event.DOC_UPLOADED);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java:251]: also that it uploaded documents from zip?

Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java:251]: that's mock definition in `given` step:

```java
willDoNothing().given(documentProcessor).uploadPdfFiles(emptyList(), emptyList());
```

if it wasn't used but stubbings defined in test, you'll get an error:

```
Unnecessary stubbings detected.
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at uk.gov.hmcts.reform.bulkscanprocessor.services.UploadEnvelopeDocumentsServiceTest.should_mark_as_uploaded_when_everything_went_well(UploadEnvelopeDocumentsServiceTest.java:243)
```

tl;dr - there's no point in verifying what you stubbed

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            .createEvent(eventCaptor.capture(), eq(CONTAINER_1), eq(ZIP_FILE_NAME), eq("oh no"), eq(null));
+        assertThat(eventCaptor.getValue()).isEqualTo(Event.DOC_UPLOAD_FAILURE);
+
+        // and
+        verify(envelopeProcessor, times(1)).markAsUploadFailure(envelopes.get(0));
+    }
+
+    @Test
+    void should_mark_as_uploaded_when_everything_went_well()
+        throws URISyntaxException, StorageException, IOException {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java:234]: just `throws Exception` would be enough

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java (hunk 1 above)
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java:251]: From readability point of view, I'd rather be able to see what the tested method is supposed to do in the 'then' stage, not in the 'given' stage.

No need to stub.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        EnvelopeProcessor envelopeProcessor
+    ) {
+        this.blobManager = blobManager;
+        this.zipFileProcessor = zipFileProcessor;
+        this.documentProcessor = documentProcessor;
+        this.envelopeProcessor = envelopeProcessor;
+    }
+
+    public void processByContainer(String containerName, List<Envelope> envelopes) {
+        log.info("Processing envelopes in {} container. Envelopes found: {}", containerName, envelopes.size());
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java:65]: I wouldn't say those envelopes were _found_ here.
Perhaps and old log from when this service was also fetching envelopes from DB?

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java (hunk 1 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java:251]: for that method call it can be converted to verification step instead of stub. 1s

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+            blobManager,
+            zipFileProcessor,
+            documentProcessor,
+            envelopeProcessor
+        );
+    }
+
+    @Test
+    void should_do_nothing_when_unknown_exception_is_caught_in_parent_try_block()
+        throws URISyntaxException, StorageException {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsServiceTest.java:72]: just `throws Exception` would be enough (here and in 9 other places)

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        EnvelopeProcessor envelopeProcessor
+    ) {
+        this.blobManager = blobManager;
+        this.zipFileProcessor = zipFileProcessor;
+        this.documentProcessor = documentProcessor;
+        this.envelopeProcessor = envelopeProcessor;
+    }
+
+    public void processByContainer(String containerName, List<Envelope> envelopes) {
+        log.info("Uploading envelope documents {} container. Envelopes count: {}", containerName, envelopes.size());
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java:65]: missing 'from' or something like that in the first sentence.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            if (lease.isPresent()) {
+                BlobInputStream bis = getBlobInputStream(blobClient, containerName, zipFileName, envelopeId);
+                ZipFileProcessingResult result = processInputStream(bis, containerName, zipFileName, envelopeId);
+
+                uploadParsedZipFileName(envelope, result.getPdfs());
+
+                envelopeProcessor.handleEvent(envelope, DOC_UPLOADED);
+            }
+        } catch (FailedUploadException failure) {
+            log.error(failure.getMessage(), failure.getCause());
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java:106]: worth logging envelope name and container at least here ?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        CloudBlockBlob blobClient = null;
+
+        try {
+            blobClient = getCloudBlockBlob(blobContainer, zipFileName, envelopeId);
+
+            Optional<String> lease = blobManager.acquireLease(blobClient, containerName, envelope.getZipFileName());
+
+            if (lease.isPresent()) {
+                BlobInputStream bis = getBlobInputStream(blobClient, containerName, zipFileName, envelopeId);
+                ZipFileProcessingResult result = processInputStream(bis, containerName, zipFileName, envelopeId);
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java:99]: is there a test case when this throws an exception ?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java (hunk 6 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java:106]: everything is logged whats possible for separate sections below

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java (hunk 7 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java:99]: the only thing that is not covered is
```
try {
    blobClient.breakLease(0);
} catch (StorageException exception) {
```
catch statement. which only logs a warning
